File: web/console/src/lib/protocol/events.ts
/**
 * Events wire types — the `events.*` Protocol shapes the Console Events
 * page consumes (Phase 73g, D-125).
 *
 * # Wire types only — the client lives in `harbor.ts`
 *
 * This module is the wire-type surface only: the request / response /
 * frame shapes the `EventsNamespace` methods (in `harbor.ts`) consume
 * and return. They mirror `internal/protocol/types/events.go` and the
 * SSE `wireEvent` projection in
 * `internal/protocol/transports/stream/frame.go` field-for-field — the
 * Go side is the single source (D-002 / D-093). When
 * `cmd/harbor-gen-protocol-ts` (D-093) ships, these absorb into the
 * generated `protocol.ts`.
 *
 * # No new Protocol method (Phase 73g)
 *
 * Phase 73g ships NO new Protocol method. `events.subscribe`
 * (`GET /v1/events` SSE — Phase 72) and `events.aggregate`
 * (`POST /v1/events/aggregate` — Phase 72a) are already shipped; the
 * truncated-payload `Open artifact` link routes through the shipped
 * `artifacts.get_ref` (Phase 73l). This page is a pure UI consumer of
 * that surface.
 *
 * # Heavy payloads flow by reference (D-026)
 *
 * The event filter operates on event HEADER fields only — type +
 * identity + timestamp. Heavy payload bytes are NEVER inlined: a
 * payload exceeding the heavy-content threshold carries an
 * {@link EventArtifactRef} the page resolves via `artifacts.get_ref`.
 *
 * # events.list (D-294) reuses the flat StateEvent row
 *
 * The `events.list` historical-window read returns the SAME flat
 * {@link StateEvent} projection `state.history` carries — imported from
 * `./state`, not re-declared (no new row shape).
 */

import type { StateEvent } from './state';

/* ------------------------------------------------------------------ */
/* EventFilter — mirrors types.EventFilter                             */
/* ------------------------------------------------------------------ */

/**
 * The canonical wire predicate every `events.*` method consumes —
 * `events.subscribe` for live subscriptions and `events.aggregate` for
 * time-bucketed count series. Mirrors `types.EventFilter`.
 *
 * Identity is mandatory. Empty `user_ids` / `session_ids` is
 * interpreted as "the caller's own identity tuple"; a `tenant_ids` set
 * with more than one entry, or a single tenant other than the caller's,
 * requires the `auth.ScopeAdmin` OR `auth.ScopeConsoleFleet` claim
 * (D-079 — there is NO dedicated `events.crosstenant` scope).
 */
export interface EventFilter {
	/** Narrows to events whose `type` is in the set. Empty matches all. */
	event_types?: string[];
	/** Tenant axis — >1 entry (or a foreign tenant) needs an admin scope. */
	tenant_ids?: string[];
	/** User axis. Empty = the caller's own user. */
	user_ids?: string[];
	/** Session axis. Empty = the caller's own session. */
	session_ids?: string[];
	/** Run axis. Empty = unconstrained. */
	run_ids?: string[];
	/** Optional inclusive lower bound on `occurred_at` (RFC-3339 UTC). */
	since?: string;
	/** Optional exclusive upper bound on `occurred_at` (RFC-3339 UTC). */
	until?: string;
}

/* ------------------------------------------------------------------ */
/* Event — mirrors the SSE wireEvent projection                        */
/* ------------------------------------------------------------------ */

/**
 * One event as it crosses the `events.subscribe` SSE wire — the flat
 * `wireEvent` projection in `frame.go`. The `payload` is whatever the
 * event's typed payload marshalled to (already redaction-safe by
 * construction — the bus runs the redactor on Publish, D-020); a heavy
 * payload carries an {@link EventArtifactRef} instead of inline bytes
 * (D-026).
 */
export interface Event {
	/** The dotted canonical event-type name (e.g. `tool.failed`). */
	type: string;
	/** The per-bus monotonic sequence — also the SSE reconnect cursor. */
	sequence: number;
	/** The wall-clock instant the event occurred (RFC-3339 UTC). */
	occurred_at: string;
	/** The publishing tenant. */
	tenant: string;
	/** The publishing user. */
	user: string;
	/** The publishing session. */
	session: string;
	/** The publishing run, when the event is run-scoped. */
	run?: string;
	/** The typed, redaction-safe payload — or an {@link EventArtifactRef}. */
	payload?: unknown;
	/** Free-form string sidecar attributes (`source`, `severity`, …). */
	extra?: Record<string, string>;
}

/**
 * A heavy-payload reference. When an event's payload exceeds the
 * heavy-content threshold (RFC §6.5 / D-026) the runtime emits the
 * payload as an `artifact_ref`-shaped object rather than inline bytes;
 * the Events page renders a `Truncated` badge + an `Open artifact` link
 * that resolves the bytes via `artifacts.get_ref`. The page NEVER
 * inlines heavy bytes into a Svelte component.
 */
export interface EventArtifactRef {
	/** Discriminator the page checks via {@link isEventArtifactRef}. */
	artifact_ref: {
		/** The artifact id the `artifacts.get_ref` resolver keys on. */
		id: string;
		/** The artifact's MIME type, when known. */
		mime?: string;
		/** The artifact's byte size, when known. */
		size?: number;
	};
}

/**
 * Narrows an event payload to an {@link EventArtifactRef}. A `true`
 * result means the page MUST route through `artifacts.get_ref` — it
 * must never read bytes off the payload directly (D-026).
 */
export function isEventArtifactRef(payload: unknown): payload is EventArtifactRef {
	return (
		typeof payload === 'object' &&
		payload !== null &&
		'artifact_ref' in payload &&
		typeof (payload as { artifact_ref: unknown }).artifact_ref === 'object' &&
		(payload as { artifact_ref: unknown }).artifact_ref !== null &&
		typeof (payload as EventArtifactRef).artifact_ref.id === 'string'
	);
}

/* ------------------------------------------------------------------ */
/* events.aggregate — mirrors types.EventAggregate{Request,Response}   */
/* ------------------------------------------------------------------ */

/**
 * One time-bucketed count series — one stripe of the per-event-type
 * stacked-area sparkline. Mirrors `types.EventBucket`. `bucket_start`
 * (inclusive) and `bucket_end` (exclusive) are RFC-3339 UTC; `counts`
 * is keyed by event-type string. Empty buckets are present (empty
 * `counts`) so the rendering client scans a contiguous time axis.
 */
export interface EventBucket {
	/** The bucket's lower bound (inclusive, RFC-3339 UTC). */
	bucket_start: string;
	/** The bucket's upper bound (exclusive, RFC-3339 UTC). */
	bucket_end: string;
	/** event-type → count of events in this bucket. */
	counts: Record<string, number>;
}

/**
 * The wire request for `events.aggregate`. Mirrors
 * `types.EventAggregateRequest`. `window` / `bucket` are Go
 * `time.Duration` nanosecond integers — `bucket` must evenly divide
 * `window` or the runtime rejects with `invalid_request`.
 */
export interface EventAggregateRequest {
	/** The event predicate the aggregator counts under. */
	filter: EventFilter;
	/** The inclusive lookback span, in nanoseconds (Go `time.Duration`). */
	window: number;
	/** The per-bucket width, in nanoseconds — must divide `window`. */
	bucket: number;
	/**
	 * OPTIONAL origin of the bucket grid (RFC-3339 UTC). Absent ⇒ the
	 * clock-anchored default (buckets run `[now-window, now)`, a
	 * `bucket_start` is NOT addressable twice). Set ⇒ boundaries are
	 * floored onto the fixed grid `anchor + k·bucket`, so two calls at
	 * two instants with the same `anchor` + `window` + `bucket` share
	 * bucket coordinates — a `bucket_start` is cacheable/re-requestable
	 * (mirrors `types.EventAggregateRequest.Anchor`, D-306). Passing the
	 * Unix epoch ({@link EPOCH_ANCHOR}) yields a globally-shared grid.
	 * Grid-edge note: with an anchor the LAST bucket's `bucket_end` is
	 * the grid boundary covering now — generally just AFTER now, never
	 * treat it as "up to this instant."
	 */
	anchor?: string;
}

/**
 * The Unix-epoch grid origin (`1970-01-01T00:00:00Z`). Passing it as
 * {@link EventAggregateRequest.anchor} floors every bucket boundary onto
 * `epoch + k·bucket`, giving a globally-shared, addressable grid the
 * Events page can cache and stitch cold windows against (D-306).
 */
export const EPOCH_ANCHOR = new Date(0).toISOString();

/** The wire response for `events.aggregate`. Mirrors `types.EventAggregateResponse`. */
export interface EventAggregateResponse {
	/** The per-bucket count series, oldest-first. */
	buckets: EventBucket[];
	/**
	 * True when the counts are PARTIAL — the aggregation could not observe
	 * every matching event in the window (a best-effort ring evicted older
	 * events below the window, or a durable log held more matching events than
	 * the single-read aggregation bound). Set uniformly across drivers; an
	 * over-wide window is DATA (partial buckets flagged truncated), never a
	 * request error. Absent/false ⇒ the counts are complete for the window.
	 */
	truncated?: boolean;
	/** The Protocol version the Runtime answered under. */
	protocol_version: string;
}

/* ------------------------------------------------------------------ */
/* events.list — mirrors types.EventsList{Request,Response}            */
/* ------------------------------------------------------------------ */

/** Default + maximum page size for `events.list`, mirroring the Go bounds. */
export const DEFAULT_EVENTS_LIST_LIMIT = 50;
export const MAX_EVENTS_LIST_LIMIT = 200;

/**
 * The wire request for `events.list` — the durable, time-ranged,
 * cross-session raw-event read. Mirrors `types.EventsListRequest`. Reuses
 * {@link EventFilter} (identity axes + `since`/`until` bounds + event-type
 * selector) plus a tail-first, sequence-based paging cursor mirroring
 * `state.history`. The `identity` triple is folded server-side; a
 * cross-tenant (fleet) filter requires the verified `admin` or
 * `console:fleet` claim (D-079). `cursor` zero ⇒ from the newest retained
 * events; pass the prior response's `next_cursor` back to scroll a page
 * older.
 */
export interface EventsListRequest {
	/** The event predicate (identity sets + since/until + types). */
	filter: EventFilter;
	/** Exclusive upper sequence bound; 0 (or omitted) ⇒ from the tail. */
	cursor?: number;
	/** Page size K (clamped to {@link MAX_EVENTS_LIST_LIMIT}; 0 ⇒ default). */
	limit?: number;
}

/**
 * The wire response for `events.list`. Mirrors
 * `types.EventsListResponse`. Rows are the SAME flat {@link StateEvent}
 * projection `state.history` and the SSE stream carry — no new row shape.
 * `next_cursor` is the value to pass back as `cursor` to scroll one page
 * older (0 at the retained head); `truncated` is the honest retention-gap
 * signal (a best-effort ring evicted older events).
 */
export interface EventsListResponse {
	/** The page's events, oldest-first within the window. */
	events: StateEvent[];
	/** Lowest sequence in this page — the scroll-up cursor (0 at the head). */
	next_cursor: number;
	/** Whether older matching events remain before this page. */
	has_more: boolean;
	/** True when a wrapped best-effort ring may have evicted older rows. */
	truncated?: boolean;
}

/* ------------------------------------------------------------------ */
/* Console-local view vocabulary                                       */
/* ------------------------------------------------------------------ */

/**
 * The active sparkline / subscription window. Console-local UI state —
 * the runtime is window-agnostic; the page derives `since` / `window` /
 * `bucket` from the selected `TimeWindow`. Mirrors page-events.md §12.
 */
export type TimeWindow = '5m' | '1h' | '24h' | '7d';

/** Nanoseconds per millisecond — the Go `time.Duration` unit bridge. */
const NS_PER_MS = 1_000_000;

/**
 * The window → (span, bucket) table page-events.md §12 pins:
 * 5 min: 30 s buckets; 1 h: 1 min; 24 h: 5 min; 7 d: 1 h. Spans and
 * buckets are expressed in nanoseconds so they feed
 * {@link EventAggregateRequest} directly.
 */
export const WINDOW_SPEC: Record<TimeWindow, { windowNs: number; bucketNs: number; label: string }> = {
	'5m': { windowNs: 5 * 60_000 * NS_PER_MS, bucketNs: 30_000 * NS_PER_MS, label: 'Last 5 min' },
	'1h': { windowNs: 60 * 60_000 * NS_PER_MS, bucketNs: 60_000 * NS_PER_MS, label: 'Last 1 h' },
	'24h': { windowNs: 24 * 60 * 60_000 * NS_PER_MS, bucketNs: 5 * 60_000 * NS_PER_MS, label: 'Last 24 h' },
	'7d': { windowNs: 7 * 24 * 60 * 60_000 * NS_PER_MS, bucketNs: 60 * 60_000 * NS_PER_MS, label: 'Last 7 d' }
};

File: internal/protocol/types/events.go
package types

import "time"

// EventFilter is the canonical wire predicate every events.* Protocol
// method consumes — `events.subscribe` for live subscriptions
// and `events.aggregate` for time-bucketed count series. The
// shape is identity-scope-aware: cross-tenant filters (TenantIDs with
// more than one entry, or a tenant other than the caller's) require the
// `auth.ScopeAdmin` OR `auth.ScopeConsoleFleet` claim per the closed admin-scope set — there
// is NO dedicated `events.crosstenant` scope (the closed two-scope set
// is the wire posture).
//
// Identity is mandatory. An EventFilter with empty UserIDs / SessionIDs
// is interpreted as "the caller's own identity tuple" — the wire edge
// resolves the missing components from the caller's identity quadruple.
// A filter that elides the triple WITHOUT a scope claim is rejected
// loudly at the wire edge with `CodeIdentityRequired`; one that names
// multiple tenants WITHOUT a scope claim is rejected with
// `CodeIdentityScopeRequired` (HTTP 403).
//
// Heavy payloads (RFC §6.5): the filter operates on event
// HEADER fields only (type, identity, timestamp). Predicates over event
// payload bytes would force the runtime to materialise heavy payloads
// through the LLM-edge safety net — explicitly out of scope per the design
// §CC-4 (substring payload search is post-V1).
//
// Since / Until are UTC; the empty zero-value means "unbounded on that
// side." A non-zero Until that precedes Since is a structurally-invalid
// filter (`CodeInvalidRequest`).
type EventFilter struct {
	// EventTypes narrows to events whose `Type` is in the set. Empty
	// matches every type (today's `events.subscribe` default). The
	// strings are the registered `events.EventType` values
	// (`tool.failed`, `planner.repair_exhausted`, etc.).
	EventTypes []string `json:"event_types,omitempty"`
	// TenantIDs / UserIDs / SessionIDs / RunIDs narrow to events whose
	// identity tuple matches one of the supplied values. Empty on any
	// axis is interpreted as "the caller's own component"; >1 on the
	// tenant axis (or a single tenant other than the caller's) requires
	// `auth.ScopeAdmin` OR `auth.ScopeConsoleFleet` per the closed admin-scope set.
	TenantIDs  []string `json:"tenant_ids,omitempty"`
	UserIDs    []string `json:"user_ids,omitempty"`
	SessionIDs []string `json:"session_ids,omitempty"`
	RunIDs     []string `json:"run_ids,omitempty"`
	// Since is the optional lower bound (inclusive) on `OccurredAt`.
	// Zero means "unbounded on the lower side." UTC.
	Since time.Time `json:"since,omitempty"`
	// Until is the optional upper bound (exclusive) on `OccurredAt`.
	// Zero means "unbounded on the upper side." UTC.
	Until time.Time `json:"until,omitempty"`
}

// Default + maximum page bounds for the `events.list` read. A client that
// omits Limit gets DefaultEventsListLimit; a Limit above MaxEventsListLimit
// is clamped down to MaxEventsListLimit (a larger ask is satisfied with the
// maximum page, not rejected). A negative Limit is the one invalid value
// (`CodeInvalidRequest`). The bounds mirror `state.history` so the two
// windowed reads share a paging grammar.
const (
	// DefaultEventsListLimit is the page size used when an `events.list`
	// request omits Limit (or sends zero).
	DefaultEventsListLimit = 50
	// MaxEventsListLimit is the largest page a single `events.list` request
	// may ask for; a Limit above it is clamped down to this value.
	MaxEventsListLimit = 200
)

// EventsListRequest is the wire request for the `events.list` Protocol
// method — the durable, time-ranged, cross-session raw-event read. It
// reuses the EventFilter above (identity axes + since/until time bounds +
// event-type selector) plus a tail-first, sequence-based paging cursor
// mirroring `state.history`'s grammar. Unlike `state.history` (by-id,
// single session) it honours the filter's multi-valued identity sets and,
// under a verified `admin` or `console:fleet` claim, fans in across
// sessions in global-sequence order. The rows returned are the SAME flat
// StateEvent projection `state.history` and the SSE stream carry — no new
// row shape, no redaction change, no heavy-payload inlining.
type EventsListRequest struct {
	// Identity is the (tenant, user, session) scope envelope every
	// Protocol client carries on the body. The handler pulls the
	// authoritative identity from the JWT / X-Harbor-* edge headers
	// (RFC §5.5) — this field is shape-only, declared so the body decode's
	// `DisallowUnknownFields` accepts the Console's standard payload shape
	// (`Transport.request` auto-injects `identity` on every request body).
	// Optional in the wire schema; ignored by the handler.
	Identity IdentityScope `json:"identity,omitempty"`
	// Filter narrows the events returned. The empty filter (zero
	// EventFilter) is interpreted as "every event in the caller's own
	// identity tuple" (per the EventFilter godoc). A cross-tenant filter
	// (a tenant other than the caller's, or multiple tenants) requires the
	// verified `admin` OR `console:fleet` claim per the closed admin-scope
	// set; scope is derived server-side, never from this body.
	Filter EventFilter `json:"filter"`
	// Cursor is the exclusive upper sequence bound (the scroll-up cursor):
	// only events with Sequence < Cursor are returned. Zero means "from the
	// tail" (the newest retained events). To scroll one page older, pass the
	// prior response's NextCursor back here.
	Cursor uint64 `json:"cursor,omitempty"`
	// Limit is the page size K (clamped to MaxEventsListLimit;
	// zero ⇒ DefaultEventsListLimit).
	Limit int `json:"limit,omitempty"`
}

// EventsListResponse is the `events.list` reply: a bounded page of flat
// events oldest-first within the window, plus a scroll-up cursor and the
// honest retention-gap signal. Row shape is byte-identical to
// `state.history`'s StateEvent projection.
type EventsListResponse struct {
	// Events is the page's events, oldest-first within the window.
	Events []StateEvent `json:"events"`
	// NextCursor is the lowest Sequence in this page — the value to pass
	// back as Cursor to scroll one page older. Zero when the retained head
	// is reached (no older events).
	NextCursor uint64 `json:"next_cursor"`
	// HasMore reports whether older matching events remain before the
	// page's oldest event.
	HasMore bool `json:"has_more"`
	// Truncated is true when the substrate's oldest retained row sits above
	// the true first matching event (a best-effort ring evicted older
	// events) — the honest gap signal, never a silent drop. False on the
	// gap-free durable log.
	Truncated bool `json:"truncated,omitempty"`
}

// EventBucket is a single time-bucketed count series — one stripe of the
// per-event-type stacked-area sparkline the Events page renders.
// Start (inclusive) and End (exclusive) are UTC; Counts is keyed
// by event-type string.
//
// Bucket boundaries are computed deterministically by the aggregator:
// for a request whose window is W and bucket size is B (B divides W),
// the response carries exactly ceil(W/B) buckets, each spanning [Start,
// Start+B). Empty buckets are present (with an empty Counts map) so a
// rendering client can scan a contiguous time axis without gap
// arithmetic.
type EventBucket struct {
	// Start is the bucket's lower bound (inclusive). UTC. The JSON tag
	// is namespaced (bucket_start rather than start) so the wire
	// surface does not collide with the task-control method
	// name start — the single-source grep backstop is
	// substring-shaped (it does not parse struct tags) and a bare
	// start tag would false-positive against it.
	Start time.Time `json:"bucket_start"`
	// End is the bucket's upper bound (exclusive). UTC. End - Start ==
	// the request's `Bucket` width.
	End time.Time `json:"bucket_end"`
	// Counts maps event-type string → count of events in this bucket.
	// An empty map means "no matching events in this window slice" —
	// the bucket is still present so the rendering client sees the gap.
	Counts map[string]int64 `json:"counts"`
	// CountsByTenant is the OPTIONAL per-tenant attribution of this
	// bucket's Counts: tenant → event-type → count. It is present ONLY
	// when the request set ByTenant AND the read was admin-widened (a
	// verified `admin` / `console:fleet` fan-in, derived server-side —
	// never from the request body). Absent (nil, omitted) on every other
	// read, so a caller that does not opt in sees a byte-identical
	// response.
	//
	// It is a pure re-projection of the SAME counted events by their
	// existing tenant identity — no new data path, no new identity axis.
	// The keys are a SUBSET of the authorized (named-or-folded)
	// Filter.TenantIDs; Counts (the totals) and CountsByTenant are scoped
	// to the IDENTICAL authorized filter by construction (one pass over
	// the same matched events), so per bucket
	// `Σ_tenant CountsByTenant[tenant][type] == Counts[type]`. That
	// reconciliation is what makes the tenant boundary independently
	// verifiable on an aggregate the way it already is on a row read
	// (every row carries its own tenant): the widened caller can check the
	// attribution keys against the entitled set it asked for.
	//
	// Empty buckets (no matching events) omit the field even under an
	// attribution request — there is nothing to attribute.
	CountsByTenant map[string]map[string]int64 `json:"counts_by_tenant,omitempty"`
}

// EventAggregateRequest is the wire request for the `events.aggregate`
// Protocol method. It returns a deterministic time series
// of event-type counts over `Window`, bucketed by `Bucket`. By default
// the window is anchored at the request's effective `Now` (the runtime's
// clock at handler entry) — the response slice runs [Now-Window, Now)
// bucketed at `Bucket` width. When the optional `Anchor` is set, the
// grid is instead floored onto `Anchor + k·Bucket` (see the field), so a
// `bucket_start` becomes a stable, addressable coordinate.
//
// The request body's `filter` is the EventFilter above; the same
// identity-scope rules apply (a cross-tenant aggregate is gated on the
// `auth.ScopeAdmin` OR `auth.ScopeConsoleFleet` claim per the closed admin-scope set).
type EventAggregateRequest struct {
	// Identity is the (tenant, user, session) scope envelope every
	// Protocol client carries on the body. The aggregate handler pulls
	// the authoritative identity from the JWT / X-Harbor-* edge headers
	// (RFC §5.5) — this field is shape-only, declared so the body
	// decode's `DisallowUnknownFields` accepts the Console's standard
	// payload shape (`Transport.request` auto-injects `identity` on
	// every request body). Optional in the wire schema; ignored by the
	// handler. Before this field existed it was the only typed request
	// shape on the Console-facing surface that REFUSED the identity
	// envelope, breaking every Events page `events.aggregate` call
	// cross-origin.
	Identity IdentityScope `json:"identity,omitempty"`
	// Filter narrows the events the aggregator counts. The empty filter
	// (zero EventFilter) is interpreted as "every event in the caller's
	// own identity tuple" (per the EventFilter godoc).
	Filter EventFilter `json:"filter"`
	// Window is the inclusive lookback span. Must be > 0. The response
	// covers [Now-Window, Now) where Now is the runtime's clock at
	// handler entry.
	Window time.Duration `json:"window"`
	// Bucket is the per-bucket width. Must be > 0 and must evenly
	// divide Window (Window % Bucket == 0) — a non-dividing pair is
	// rejected with `CodeInvalidRequest` so the response's bucket count
	// is deterministic and a rendering client never sees a fractional
	// trailing bucket.
	Bucket time.Duration `json:"bucket"`
	// Anchor is the OPTIONAL origin of the bucket grid. When nil (the
	// default), the grid is anchored at the effective Now — the
	// response covers [Now-Window, Now) and a `bucket_start` is NOT
	// addressable twice (two calls at two instants return two boundary
	// sets). When set, bucket boundaries are floored onto the fixed grid
	// `Anchor + k·Bucket`: the response still covers a `Window`'s worth
	// of buckets, but aligned to that grid, so two calls at two instants
	// with the SAME Anchor + Window + Bucket share boundary instants — a
	// `bucket_start` is a stable coordinate a consumer can cache and
	// re-request, INCLUDING across a runtime restart and across two
	// fleet runtimes (the boundary is a pure function of Anchor+Bucket,
	// with no residual dependence on Now). Passing the Unix epoch
	// (`1970-01-01T00:00:00Z`) yields a globally-shared grid.
	//
	// It is a POINTER (not a `time.Time` value) because `omitempty` is a
	// no-op on a struct — a value type would always serialize the zero
	// time and break the additive/optional contract and its TS mirror.
	//
	// Grid-edge note: with an Anchor, the LAST bucket's `bucket_end` is
	// the grid boundary that covers Now — it is generally AFTER Now (by
	// up to one Bucket width), not equal to it. A rendering client must
	// therefore not treat the final bucket as "up to this instant"; the
	// grid re-phases the boundaries and never widens the read span by
	// more than one Bucket at either edge. UTC.
	Anchor *time.Time `json:"anchor,omitempty"`
	// ByTenant opts IN to per-tenant attribution on each bucket
	// (EventBucket.CountsByTenant). It is HONOURED ONLY on an
	// admin-widened read (a verified `admin` / `console:fleet` cross-
	// principal-or-multi-value fan-in, derived server-side); on any other
	// read it is IGNORED and the response is byte-identical to a request
	// that never set it (fail-closed — the body flag never elevates a
	// read, and never widens what is counted).
	//
	// When honoured, each bucket carries CountsByTenant re-projecting that
	// bucket's Counts by the counted events' existing tenant identity, so
	// the widened caller can independently verify the tenant boundary the
	// runtime enforced (the attribution keys must be a subset of the
	// tenants it asked for, and they must sum back to the totals). It adds
	// NO new identity axis and NO payload — a count per `(tenant,
	// event_type)` is the whole surface. Absent/false ⇒ no attribution.
	ByTenant bool `json:"by_tenant,omitempty"`
}

// EventAggregateResponse is the wire response for the `events.aggregate`
// Protocol method. Buckets are in chronological order (oldest first);
// `len(Buckets) == Window/Bucket`.
type EventAggregateResponse struct {
	// Buckets is the per-bucket count series, oldest first. Each
	// bucket's [Start, End) span is exactly `Request.Bucket` wide. When
	// the request carried no `Anchor`, the last bucket's End equals the
	// request's effective Now; when it carried an `Anchor`, the last
	// bucket's End is the grid boundary covering Now (generally AFTER
	// Now, by up to one Bucket — see EventAggregateRequest.Anchor).
	Buckets []EventBucket `json:"buckets"`
	// Truncated is true when the counts are PARTIAL: the aggregation could
	// not observe every matching event in the window. It is set UNIFORMLY
	// across drivers — a best-effort ring that has evicted matching events
	// (older matches may be lost) sets it, and a durable log whose window held more
	// matching events than the single-read aggregation bound could count
	// sets it too. It is the honest DATA signal that a driver difference
	// (retention depth, an observed horizon) changed WHAT the method
	// returns without ever changing WHETHER it works: an over-wide window is
	// never a request error and never a silent undercount — it is partial
	// buckets flagged truncated. False means the counts are complete for the
	// window.
	Truncated bool `json:"truncated,omitempty"`
	// ProtocolVersion echoes the Protocol version the Runtime answered
	// under so a client can detect a version skew (mirrors the
	// ControlResponse / StartResponse shape).
	ProtocolVersion string `json:"protocol_version"`
}
